packages/cli/src/project-references.ts

```diff
@@ -0,0 +1,184 @@
+   *
+   * @param node - The node to visit.
+   */
+  function visit(node: Value) {
+    if (recursionStack.has(node)) {
+      cycle = true;
+      return;
+    }
+
+    if (visited.has(node)) {
+      return;
+    }
+
+    recursionStack.add(node);
+    visited.add(node);
+
+    const neighbours = graph.get(node);
+    assert(neighbours !== undefined);
+
+    neighbours.forEach(visit);
+    recursionStack.delete(node);
+    stack.push(node);
+  }
+
+  for (const node of graph.keys()) {
+    visit(node);
+  }
+
+  return {
+    stack,
+    cycle,
+  };
+}
+
+/**
+ * Get the resolved project references from a TypeScript program.
+ *
+ * @param resolvedProjectReferences - The resolved project references of the
+ * package that is being built.
+ * @returns The resolved project references.
+ */
+export function getResolvedProjectReferences(
+  resolvedProjectReferences: ResolvedProjectReference[],
+) {
+  const graph = createGraph(resolvedProjectReferences);
+  const { stack, cycle } = topologicalSort(graph);
+
+  if (cycle) {
+    throw new Error(
+      'Unable to build project references due to a dependency cycle.',
+    );
+  }
+
+  return stack;
+}
+
+/**
+ * Get a list of the output file paths in the referenced projects.
+ *
+ * @param resolvedProjectReferences - The resolved project references of the
+ * package that is being built.
+ * @returns A list of output paths.
+ */
+export function getReferencedProjectPaths(
+  resolvedProjectReferences: readonly ResolvedProjectReference[],
+): string[] {
+  return resolvedProjectReferences.flatMap(({ commandLine }) => {
+    return commandLine.fileNames.flatMap((fileName) =>
+      getOutputFileNames(commandLine, fileName, false),
+    );
+  });
+}
+
+/**
+ * Create a compiler host that can be used to build projects using
+ * project references.
+ *
+ * This is almost the same as the default compiler host, but it modifies the
+ * `getSourceFile` method to look at `.d.cts` files instead of `.d.ts` files for
+ * declaration files.
+ *
+ * @param compilerOptions - The compiler options to use.
+ * @param resolvedProjectReferences - The resolved project references of the
+ * package that is being built.
+ * @returns The compiler host.
+ */
+export function createProjectReferencesCompilerHost(
+  compilerOptions: CompilerOptions,
+  resolvedProjectReferences: readonly ResolvedProjectReference[],
+): CompilerHost {
+  const host = createCompilerHost(compilerOptions);
+  const originalGetSourceFile = host.getSourceFile.bind(host);
+  const referencedProjectPaths = getReferencedProjectPaths(
+    resolvedProjectReferences,
+  );
+
+  const getSourceFile: CompilerHost['getSourceFile'] = (fileName, ...args) => {
+    if (!referencedProjectPaths.includes(fileName)) {
+      return originalGetSourceFile(fileName, ...args);
+    }
+
+    // TypeScript checks the referenced distribution files to see if the project
+    // is built. We simply point it to the `.d.cts` files instead of the `.d.ts`
```

Comment:
Does tsc not support `d.cts` out of the box?

Reply:
It does, but only if the source file is `.cts`. We use `.ts` files and TypeScript assumes these to be built as `.d.ts`, but since TS Bridge modifies the output paths, we need to point TypeScript to the `.cts` files.

Reply:
How does `.d.mts` figure into this?

Reply:
I changed this later to use `sourceExtension` (`.mts` for ES modules, `.cts` for CommonJS), which is fetched from the first `format`. So when building with `--format module` or `--format module commonjs, it will redirect TypeScript to the `.d.mts` instead.

Reply:
Ah, I see. The comment is misleading. 👍